concepts/iterators/introduction.md

```diff
@@ -0,0 +1,173 @@
+- And any type that implements `Iterator` is an iterator, which _produces_ values and can be _iterated over_.
+- You can get an iterator from any type that implements `IntoIterator` by calling its `into_iter` method.
+
+## `into_iter` and for loops
+
+As you can see, any type that implements `IntoIterator` has an `into_iter` method.
+In fact, you're probably familiar with for loops over a vector (`Vec<T>`):
+
+```rust
+// capitals is a `Vec`, which implements `IntoIterator`
+let lor = vec!["Ainur", "Elves", "Men", "Dwarves", "Hobbits", "Orcs", "Dragons"];
+for people in &lor {
+    println!("{}", people);
+}
+```
+
+which are just a shorthand for calling the vector's `into_iter` method to turn the `iterable` into an `iterator`
+and then calls the `next` method on the `iterator` until it produces no more values (i.e. returns `None`):
+
+```rust
+// capitals is a `Vec`, which implements `IntoIterator`
+let mut peoples = (&lor).into_iter();
+
+// capitals_iterator implements `Iterator`, which has the default `next` method
+while let Some(people) = peoples.next() {
+    println!("{}", people);
+}
+```
+
+## `iter` and `iter_mut` methods
+
+Most Rust collection types (such as `Vec`, `HashMap`, `HashSet`) provide `iter` and `iter_mut` methods.
+For an iterator that produces values of type `T`, these two methods perform the following by convention:
+
+- The iterator returned by `iter` will yield immutably borrowed `&T`.
+- The iterator returned by `iter_mut` will yield mutably borrowed `&mut T`.
+
+```rust
+let mut v = vec![1, 2, 3];
+
+// Using `iter` method
+let mut iterator = v.iter();
+assert_eq!(iterator.next(), Some(&1));
+assert_eq!(iterator.next(), Some(&2));
+assert_eq!(iterator.next(), Some(&3));
+assert_eq!(iterator.next(), None);
+
+// Using `iter_mut` method
+let mut muterator = v.iter_mut();
+assert_eq!(muterator.next(), Some(&mut 1));
+assert_eq!(muterator.next(), Some(&mut 2));
+assert_eq!(muterator.next(), Some(&mut 3));
+assert_eq!(muterator.next(), None);
+```
+
+## Difference between `iter`, `iter_mut`, and `into_iter`
+
+Now that we've covered `iter`, `iter_mut`, and `into_iter`, what are the differences?
+Let's amend our previous bullet points:
+
+- The iterator returned by `iter` will yield immutably borrowed `&T`.
+- The iterator returned by `iter_mut` will yield mutably borrowed `&mut T`.
+- The iterator returned by `into_iter` may yield `T`, `&T` or `&mut T`, depending on its context.
+
+Wait a second, what do we mean by _depending on its context_?
+Just what `into_iter` is invoked on.
+Let's clarify using the following bullet points:
+
+- `(&v).into_iter()` is equal to `v.iter()` => returns `&T`
+- `(&mut v).into_iter()` is equal to `v.iter_mut()` => returns `&mut T`
+- `v.into_iter()` => returns `T` and moves the values
+
+This is most clear in for loops:
+
+- `for num in &v {...}` => returns `&T`
+- `for num in &mut v {...}` => returns `&mut T`
+- `for num in v {...}` => returns `T` and moves the values
```

Comment:
```suggestion
- `for num in &v {...}` => binds `&T`
- `for num in &mut v {...}` => binds `&mut T`
- `for num in v {...}` => binds `T` and moves the values
```

We're binding the variable name `num` in each of these cases, not returning anything.